Submit application of hamster (#789)
w3f/Grants-Program: applications/hamster.md

Our team aims to build a complete platform for leasing idle computing resources. The decentralized computing resource leasing function is realized by using substrate framework pallet, which can be leased and traded on the chain, and the link and use of computing resources can be done through the p2p network protocol under the chain after signing the lease agreement on the chain.

### Project Details

Hamster is composed of Hamster nodes, Hamster resource providers, Hamster clients, Hamster gateways, and Hamster front-end pages.

![](https://gitee.com/ltyuanmu/drawing-bed/raw/master/202201121240597.jpg)

- Hamster Node: is a custom node built on Substrate 3.0.
  - pallet_provider: performs resource provider registration and provider resource information storage, and provides a computational marketplace
  - pallet_resource-order: performs resource lease order functions and lease agreement execution
  - pallet_gateway: The gateway mainly has the following functions: gateway registration, gateway heartbeat detection, gateway status reset, gateway drop penalty, and receive rewards. The main role is to add the gateway as an important player in the shared computing platform

> - Hamster Gateway: p2p gateway with public IP, used to link information between resource provider and resource user, built with libp2p component, is the cornerstone of the leased resource availability, can register itself to Hamster Node. That include **Register**,**Receive rewards ** and **Configuration** module.

Reviewer: ```suggestion
- Hamster Gateway: p2p gateway with public IP, used to link information between resource provider and resource user, built with libp2p component, is the cornerstone of the leased resource availability, can register itself to Hamster Node. That includes the **Register**, **Receive rewards** and **Configuration** modules.
```

  - Register: Register Gateway information to Hamster Node.
  - Receive rewards: Revenue for those who provide gateway resources.
  - Configuration: Basic configuration of p2p gateway information.

> - Hamster Provider: can provide compute resources and register them with Hamster Node. Compute resources are provided using both vm virtual machine technology and docker technology. Currently vm virtual machine technology is used to better protect user privacy.That include **Initialize configuration**,**Resource details**,**Account information** and **Configuration information** module.

Reviewer: ```suggestion
- Hamster Provider: can provide compute resources and register them with Hamster Node. Compute resources are provided using both vm virtual machine technology and docker technology. Currently vm virtual machine technology is used to better protect user privacy. That includes the **Initialize configuration**, **Resource details**, **Account information** and **Configuration information** modules.
```

  - Initialize configuration: Initialize configuration, including p2p seed node configuration, p2p port configuration.
  - Resource details: Available spare resource specifications (cpu, memory, etc.), and price.
  - Account information: Provide the import of the substrate account, provide the pledge of the deposit before the service, etc.
  - Configuration information: resources that have not reached a transaction are offline at any time, resource specifications, price configuration, etc.

> - Hamster Client: After purchasing in the front end marketplace, users can view their purchased compute resources through the client and link to them. That include **Market**,**My order** and **My resource** module.

Reviewer: ```suggestion
- Hamster Client: After purchasing compute resources in the front-end marketplace, users can view them through the client and link to them. That includes the **Market**, **My order** and **My resource** modules.
```

  When the agreement is reached, the provider is required to complete the creation of the virtual machine, notify the blockchain of the completion of the order agreement, and request a heartbeat per epoch, which includes the number of cpu cores and memory size, for the blockchain to verify that it is working and providing valid resources. The potential threat is a nefarious provider forging heartbeats to trick the network into providing a valid service that actually provides less than the promised resources or even none at all. Since we have thought long and hard about this, and we do not currently have a cryptographic or better design to accomplish such a solution, we have adopted the idea of an arbitration model to prevent malicious attacks.

  ![](https://gitee.com/ltyuanmu/drawing-bed/raw/master/202201241409500.jpg)

  1. when the user finds a problem with the resource and the information configured at the time of purchase does not match, submits a validation request and requests arbitration. in order to ensure that there is no malicious validation, the submission of the request will have a certain cost pledge.

>   2. the arbitration group is subject to the validation request, currently designed to arbitrate the group for the current chain of validators. because he is the block validator, will promote the role of the current ecology. It is a trusted team

Reviewer: > arbitration group

You mention "arbitration group" multiple times, however, the graphic above contains only an "arbitration college". I guess you use these terms interchangeably?! If that's the case, it would be good if only 1 of these terms would be used to avoid confusion. Otherwise, could you please explain the difference?

Author: Sorry we used the wrong words in the graphic, we have fixed it

- https://github.com/hamster-shared/hamster-provider

- https://github.com/ltmuyuan
- https://github.com/mohaijiang
- https://github.com/jian-guo-s
- https://github.com/LwwL-123

## Development Status :open_book:

We have completed a simple version of the project for minimal process execution, and will subsequently update the usage documentation so you can install and experience using it.

The following is a list of the features that have been implemented:

- Hamster Node

  Computer resource register

  Update computer resource unit-price

  Create resource order

  Provider finish order

  Provider heartbeat

  Provider staking amount

  Provider withdraw amount

  Withdraw rental amount

  Withdraw fault execution

  Client cancel order

- Hamster Provider

  Computer create

  Computer destruction

  P2p link listen

- Hamster Client

  List of paid computers

  Computer link usage

- Hamster Front End

  Computing market

  Provider shared computing

  Provider lease agreement list

  List of computers purchased by the customer

  Customer order list

- Hamster Gateway

  Private IPFs network

## Development Roadmap :nut_and_bolt:

### Overview

> - **Total Estimated Duration:** ~~3 months~~ 2months

Reviewer: ```suggestion
- **Total Estimated Duration:** 2 months
```

- **Full-Time Equivalent (FTE):**  4 FTE

> - **Total Costs:** ~~$38,400~~ $25,600

Reviewer: ```suggestion
- **Total Costs:** $25,600
```

### Milestone 1 Example — Implement Hamster Client and Provider Modules

- **Estimated duration:** 1 month
- **FTE:**  4
- **Costs:** $12,800

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| 0a. | License | Apache 2.0 |
| 0b. | Documentation | We will provide online documentation of the code and a basic tutorial that includes<br/>1. Hamster Client installation tutorial<br/>2. Hamster Client usage tutorial<br/>3. Hamster Provider installation tutorial<br/>4. Hamster Provider usage tutorial<br/>5. Hamster Chain installation tutorial<br/>6. Hamster Chain usage tutorial |
| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |

> | 1. | Hamster Client | We will create a desktop client that will have **Market**,**My order** and **My resource** module.<br/>We will provide a desktop (windows, macos) app based on [wails](https://wails.io/) to achieve<br/>1: Market inquiry and management of purchased resources,<br/>2: The point-to-point communication between the client and the computing resource provider.<br/>3: Query historical orders<br/>4: Partial fee refund for defaulted orders<br/>Tech stack: go+wails+go-libp2p+vue.js+polkadot.js |

Reviewer 2: Could you deliver a Linux version of the desktop client? Note that all software delivered as part of a grant [must not rely on closed-source software](https://github.com/w3f/Grants-Program#guidelines).

Author: We use Wails is an open source go-based development client framework, we currently use his V2 version is to support windows and macos, we can provide subsequent Linux version of the desktop client

Reviewer 2: Fair enough, I see Wails v2 will eventually support Linux too. Thanks.

> | 2. | Hamster Provider | We will create a a resource provider server that will have **Initialize configuration**,**Resource details**,**Account information** and **Configuration information** module.<br/>The resource provider will use the idle resources of the machine, register in the market, and declare that it can provide rental. When the provided virtual machine reaches a transaction in the market, the resource provider provides the virtual machine with the corresponding quota for the remote purchaser to use according to the agreement. The resource provider will establish a p2p connection with the used client for remote management, such as ssh, rdesktop, etc.<br/>Our initial idea is to use the libvirt scheme to implement the management of virtual machines. (Windows uses Hyper-V)<br/>Tech stack: go+go-libp2p+libvirt+go-substrate-rpc-client+gin |

Reviewer: > Our initial idea is to use the libvirt scheme to implement the management of virtual machines. (Windows uses Hyper-V)

Does that mean, Windows will not be supported? Can you maybe specify the operating systems that will be supported?

Author: I apologize for the late reply, I have fixed the problem with the document, thank you very much.

| 4. | Hamster chain | pallet_provider, pallet_resource-order modifications and optimizations due to the need to optimize and adapt already developed pallets when adding features for customers. As code delivery of the underlying framework, there are two integrated pallets |
| 5. | Whitepaper | Preparation of project white papers |

### Milestone 2 Example — Implement Hamster Gateway Modules

- **Estimated duration:** 1 month
- **FTE:**  4
- **Costs:** $12,800

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| 0a. | License | Apache 2.0 |
| 0b. | Documentation | We will provide online documentation of the code and a basic tutorial that includes<br/>1. Hamster Gateway installation tutorial<br/>2. Hamster Gateway usage tutorial |
| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
| 0d. | Article | We will write an article or tutorial that explains the work done as part of the grant. |

> | 1. | Hamster Gateway | We will create a desktop client that will have **Register**,**Receive rewards ** and **Configuration** module.<br/>Tech stack: go+go-libp2p+libvirt+go-substrate-rpc-client+gin |

Reviewer: ```suggestion
| 1. | Hamster Gateway | We will create a desktop client that will have **Register**, **Receive rewards** and **Configuration** module.<br/>Tech stack: go+go-libp2p+libvirt+go-substrate-rpc-client+gin |
```

| 2. | Hamster Node: pallet_gateway | We will create a Substrate module that will have **Register gateway**,**Receive rewards** etc function. |
| 3. | Hamster Gateway: web app | We will create a web app integrated in the Hamster Gateway, to let users easily interact with our Hamster Gateway function module.<br/>Tech stack: vue.js+polkadot.js+Node.js |
| 4. | Hamster chain | Add a gateway module in the Hamster chain, to enable users to share gateway resources with public IP , receive rewards, etc. |

> 
> ### ~~Milestone 3 Example — Implement Hamster Anti-malicious attacks Modules~~
> 
> - ~~**Estimated duration:** 1 month~~
> - ~~**FTE:**  4~~
> - ~~**Costs:** $12,800~~
> 
> | ~~Number~~ | ~~Deliverable~~ | ~~Specification~~ |
> |-------:| ----------- | ------------- |
> | ~~0a.~~ | ~~License~~ | ~~Apache 2.0~~ |
> | ~~0b.~~ | ~~Documentation~~ | ~~We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can use Hamster to share and use computing resources~~ |
> | ~~0c.~~ | ~~Testing Guide~~ | ~~Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests.~~ |
> | ~~0d.~~ | ~~Article~~ | ~~We will write an article or tutorial that explains the work done as part of the grant.~~ |
> | ~~1.~~ | ~~Evil punishment~~ | ~~Design and implementation of punishment module for evil doing~~ |
> | ~~2.~~ | ~~Hamster Front end~~ | ~~Front end function addition and optimization to let users easily interact with our Hamster function module.~~ |
> | ~~3.~~ | ~~Benchmark~~ | ~~Perform unit tests on the individual algorithms to ensure system safety.~~ |

Reviewer: ```suggestion
```

## Future Plans

> Our current goal is to provide distributed shared computing power. After this milestone is completed, we will be able to build our own ecosystem based on resources. That is, when everyone is willing to add their free resources to the hamster network, it can be assumed that there are near-infinite computing resources in the network, and more quality services can be built and constructed in the hamster network. For example, service providers no longer need to rely on the support of a single cloud vendor, but only need to use the computing resources in the hamster network, and the system automatically schedules the resource arithmetic needed for the services to build their own services in the form of edge computing. In addition to this, we can also provide some toolkits so that these computing resources can participate in other services with one click, such as becoming a chain node of Thegraph, so that they can participate in other networks, and later on, we can also use incentives to encourage people to develop toolkits for different ecologies. When more and more people and more service providers participate in the hamster network, maybe it can become a truly decentralized meta-universe service cornerstone.

Reviewer: > we can also use incentives to encourage people to develop toolkits for different ecologies.

What do you mean by "different ecologies"? Or did you mean "different technologies" instead?

> maybe it can become a truly decentralized meta-universe service cornerstone.

I'm not sure what that means.

Author: Sorry, we may not have made it clear that "different ecologies" here means that we can provide computing power to other projects in the blockchain ecosystem that need computing resources. And we hope to support the development of plugins that allow our computing resources to provide services directly, so that real users can quickly support their favorite blockchain projects.

Regarding the cornerstone of the meta-verse service, in Chinese we call it the infrastructure of the meta-verse. I think the meta-verse is virtualized in our world by consuming a lot of computational resources to accomplish this. Computational resources are the basis for carrying the meta-verse computation, and I think maybe the computational nodes of the meta-verse are mega arithmetic centers and supercomputer nodes etc. for providing. Perhaps I have the luxury of hoping that our computing services will reach a certain scale and we can provide our computing power in the meta-verse as well.

Author: Yes, you're right, that's what we're trying to say, thank you very much
